crates/bevy_ecs/src/observer/trigger_event.rs

```diff
@@ -100,88 +101,88 @@ fn trigger_event<E: Event, Targets: TriggerTargets>(
 /// [`Observer`]: crate::observer::Observer
 pub trait TriggerTargets {
     /// The components the trigger should target.
-    fn components(&self) -> &[ComponentId];
+    fn components(&self) -> impl Iterator<Item = ComponentId> + Clone;
 
     /// The entities the trigger should target.
-    fn entities(&self) -> &[Entity];
+    fn entities(&self) -> impl Iterator<Item = Entity>;
 }
```

Comment:
Worth noting that this effectively reverts the change to this trait from https://github.com/bevyengine/bevy/pull/13991 which made it object-safe. I don't see any discussion for that original change, but if it's being changed back, it probably warrants some. At the very least, we should probably go back to the original `impl ExactSizeIterator<...>` return types if possible to keep it consistent with `0.14.0` and minimize potential for breakage.

Reply:
Yes I was hoping to see more discussions on the consequences of this change; I wasn't sure why the change was made in the first place.

Note that `impl ExactSizeIterator` is unfortunately not possible because calling `chain` on 2 `ExactSizeIterator` does not return an `ExactSizeIterator`

Reply:
We could also just provide our own `Iterator` type to make this object safe

Reply:
After digging into #14775, returning Iterators would simplify its implementation. So I'm in favor of this change.